<a href="https://colab.research.google.com/github/Patortiz26/Laboratorio-de-Programaci-n/blob/main/Laboratorio4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1><center>Laboratorio 4: El Pandas no muerde (act. II) 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesor: Pablo Badilla
- Auxiliar: Ignacio Meza D.
- Ayudante: Constanza Peña

### Equipo: 

- Nombre de alumno 1: Patricio Ortiz

### **Link de repositorio de GitHub:** `http://....`

# Temas a tratar

- Manejo de datos tabulares usando `pandas`. En esta segunda parte se incluye adicionalmente agregaciones, concatenaciones, merge y trabajo con strings.
- Visualizaciones interactivas de los datos con `plotly`.

### Reglas:

- Fecha de entrega: 8 de octubre (atrasos hasta el 11 de octubre, 1 punto de descuento por día)
- **Grupos de máximo 2 personas**
- **Ausentes** deberán realizar la actividad solos. 
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias. 
- Pueden usar cualquer matrial del curso que estimen conveniente.


### Objetivos principales del laboratorio

- Aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas`.
- Utilizar `plotly` para obtener información gráfica del dataset.
- Aplicar el **Análisis Exploratorio de Datos** a un caso en particular.


> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

# Importamos librerias utiles 😸

In [35]:
# Libreria Core del lab.
import numpy as np
import pandas as pd 
import os

from IPython.display import display

#Libreria para visualizar
#!pip install --upgrade plotly
import plotly.figure_factory as ff
import plotly.express as px
import plotly.graph_objects as go

# 1. Rendimiento en Estudiantes 📚


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

Para este laboratorio deberán continuar el Análisis Exploratorio de datos sobre el conjunto ```students_grades.csv```, el cual contiene una caracterización sobre el rendimiento y otros atributos de cada alumno de la *Universidad de la Cachaña* .

In [22]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = '/content/drive/My Drive/Colab Notebooks/MDS7202 - Laboratorio de programación/Lab 4/Datos'
    os.chdir(path)
except: 
    print('Ignorando conexión drive-colab')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# 1. Carga de Datos [0.5 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente). 

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: ´students_grades_1.json´ y ´students_grades_2.json´. 


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos. No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas, recuerde que anteriormente ya se encargo de este tema.

In [26]:
df1 = pd.read_json('students_grades_1.json')
df2 = pd.read_json('students_grades_2.json')

In [27]:
df1.head()

names  gender  ... reading score writing score
0     Rita Courtney  female  ...          3.76          3.76
1  Charles Linstrom    male  ...          5.68          5.86
2       Brian Young    male  ...          4.96          4.78
3    Howard Jimenez    male  ...          5.50          5.56
4      Wayne Wilson    male  ...          6.16          6.22

[5 rows x 9 columns]

In [28]:
df2.head()

names  gender  ... reading score writing score
0     Cheryl Kraus  female  ...          5.32          5.80
1     Kristy Glass  female  ...          5.44          5.38
2  Timothy Botello    male  ...          6.04          6.34
3     Ronald Coria    male  ...          6.52          6.16
4   Jessica Martin  female  ...          5.20          5.20

[5 rows x 9 columns]

In [173]:
df_grades = pd.concat([df1,df2],axis=0)
df_grades

names  gender  ... reading score writing score
0       Rita Courtney  female  ...          3.76          3.76
1    Charles Linstrom    male  ...          5.68          5.86
2         Brian Young    male  ...          4.96          4.78
3      Howard Jimenez    male  ...          5.50          5.56
4        Wayne Wilson    male  ...          6.16          6.22
..                ...     ...  ...           ...           ...
470     Richard Young    male  ...          5.50          5.26
471     Wanda Russell  female  ...          3.64          3.16
472    Marina Zeigler  female  ...          5.44          5.86
473     Laurie Carter  female  ...          4.66          4.72
474      Amanda Perez  female  ...          5.80          5.56

[875 rows x 9 columns]

Resultado esperado:

In [ ]:
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
470     Richard Young    male        group D                 high school   
471     Wanda Russell  female        group B                 high school   
472    Marina Zeigler  female        group C           bachelor's degree   
473     Laurie Carter  female        group B            some high school   
474      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
470      standard                    none        5.14           5.50   
471  free/reduced               completed        2.38           3.64   
472  free/reduced               completed        4.96           5.44   
473      standard               completed        4.24           4.66   
474      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
470           5.26  
471           3.16  
472           5.86  
473           4.72  
474           5.56  

[875 rows x 9 columns]

## Análisis de Las Notas v2 [1.5 Puntos]

Preocupado por la dificultad que representan los boxplot, el rector le solicita buscar alternativas de visualización la distribución de las notas. 

Para esto, muestre el [boxplot](https://plotly.com/python/box-plots/) mostrado la clase anterior, utilice un [displot](https://plotly.com/python/distplot/#distplot-with-pandas) como también un [histograma con un gráfico marginal de caja](https://plotly.com/python/histograms/#visualizing-the-distribution) como también un [histograma con el ramo como faceta de fila](https://plotly.com/python/facet-plots/#histogram-facet-grids).

Luego, responda las siguientes pregunta: 

> 1. ¿Existe una diferencia notable entre las notas?

> 2. ¿Cuál de los gráficos mostrados cree que es adecuado para mostrarle al rector? ¿Y a los padres? ¿Y a un centro de estudios educativos? ¿Por qué?. Base sus respuestas en lo visto en la clase de visualizaciones como también en lo que usted y su equipo consideren correcto.

> Hint: Para elaborar el histograma, puede que le sea de utilidad hacer un `melt` del DataFrame, dejando como variables los ramos y valores las notas. Por otra parte, visiten la documentación para generar los gráficos.


**Gráfico de Caja:**

In [ ]:
def plot_box(df):
  fig = go.Figure()
  fig.add_trace(go.Box(y=df['math score'],name='Math'))
  fig.add_trace(go.Box(y=df['reading score'],name='Reading'))
  fig.add_trace(go.Box(y=df['writing score'],name='Writing'))
  fig.update_xaxes(title_text="Test")
  fig.update_yaxes(title_text="Score")

  fig.show()

plot_box(df_grades)

**Distplot:**

In [ ]:
def plot_dist(df):
  hist_data = [df['math score'],df['reading score'],df['writing score']]
  labels = ['math score','reading score','writing score']

  fig = ff.create_distplot(hist_data, labels, show_hist = False, bin_size=.25, show_rug=False)
  fig.show()  

plot_dist(df_grades)

**Histograma con Boxplots:**

In [ ]:
def plot_hist_box(df):
  math = df[['math score']]
  math.columns = ['score']
  math['subject'] = 'math score'

  read = df[['reading score']]
  read.columns = ['score']
  read['subject'] = 'reading score'

  write = df[['writing score']]
  write.columns = ['score']
  write['subject'] = 'writing score'

  qw = pd.concat([math,read,write],axis=0)

  fig = px.histogram(qw, x='score', color='subject',
                   marginal="box", nbins=7, barmode='group')
  fig.show()

plot_hist_box(df_grades)

**Histograma con Faceta:**

In [ ]:
def plot_hist_face(df):
  math = df[['math score']]
  math.columns = ['score']
  math['subject'] = 'math score'

  read = df[['reading score']]
  read.columns = ['score']
  read['subject'] = 'reading score'

  write = df[['writing score']]
  write.columns = ['score']
  write['subject'] = 'writing score'

  qw = pd.concat([math,read,write],axis=0)

  fig = px.histogram(qw, x='score', color='subject',
                   facet_row='subject')
  fig.show()

plot_hist_face(df_grades)

**Justifique:**

¿Existe una diferencia notable entre las notas?

¿Cuál de los gráficos mostrados cree que es adecuado para mostrarle al rector? ¿Y a los padres? ¿Y a un centro de estudios educativos? ¿Por qué?. Base sus respuestas en lo visto en la clase de visualizaciones como también en lo que usted y su equipo consideren correcto

> No, no existe una diferencia notable entre las notas en las 3 asignaturas.

> A mi juicio, la visualización de histograma con boxplots es la más adecuada de presentar. Pues es la visualización que mezcla la simplicidad para generar entendimiento en el espectador a su vez que incorpora la mayor cantidad de información, en palabras sencillas esta visualización "tiene mucha información y es fácil de entender y explicar". Entre los beneficios de esta visualización se puede mencionar que es simple de identificar la distribución de las notas "por segmento de nota" y la distribución general de todas las notas.

## Análisis por Nivel Educacional Etnia de los Padres [1.5 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos. 
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.

### Tabla de Resumen

Para generar la tabla de resumen:

- [x] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [x] Hacer una simplificación a través de un mapeo (investigar `map()`) de la variable `parental level of education`  según la siguiente conversión: 

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [x] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas. 
- [x] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [x] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'. 


Utilizar la tabla de resultados esperados como guía para desarrollar este punto.

### 1.3.4 Información por Categorías [1 punto]

In [178]:
def estudia_hipotesis(df):
  df['GPA'] = np.round((df['math score'] + df['reading score'] + df['writing score'])/3,2)

  df['simple parental level of education'] = df['parental level of education'].map(
    {   "some high school": "school",
        "some college": "school",
        "high school": "school",
        "bachelor's degree": "college",
        "associate's degree": "college",
        "master's degree": "postgraduate"    })
  
  qw = df.groupby(['race/ethnicity','simple parental level of education']).agg({'math score':['mean'],
                                                                         'reading score':['mean'],
                                                                         'writing score':['mean'],
                                                                         'GPA':['mean','count']})
  #qw = qw.reset_index()
  qw.columns = ['math score', 'reading score', 'writing score', 'GPA', 'count']

  total = len(df)
  porcentaje = [100*i/total for i in qw['count']]
  porcentaje = [str(i)[:4]+'%' for i in porcentaje]
  qw['porcentaje'] = porcentaje
  qw = qw.reset_index(level='simple parental level of education')

  return qw

df_grades_new = estudia_hipotesis(df_grades)
df_grades_new

simple parental level of education  math score  ...  count  porcentaje
race/ethnicity                                                 ...                   
group A                                   college    4.735000  ...     24       2.74%
group A                              postgraduate    4.690000  ...      2       0.22%
group A                                    school    4.568235  ...     51       5.82%
group B                                   college    5.070000  ...     54       6.17%
group B                              postgraduate    4.912000  ...      5       0.57%
group B                                    school    4.689159  ...    107       12.2%
group C                                   college    5.021765  ...    102       11.6%
group C                              postgraduate    4.924000  ...     15       1.71%
group C                                    school    4.764516  ...    155       17.7%
group D                                   college    5.108286  ...     70        8.0%
group D                              postgraduate    5.218000  ...     20       2.28%
group D                                    school    5.018792  ...    149       17.0%
group E                                   college    5.542692  ...     52       5.94%
group E                              postgraduate    5.540000  ...      6       0.68%
group E                                    school    5.396190  ...     63        7.2%

[15 rows x 7 columns]

In [175]:
df_grades_new

race/ethnicity simple parental level of education  ...  count  porcentaje
0         group A                            college  ...     24       2.74%
1         group A                       postgraduate  ...      2       0.22%
2         group A                             school  ...     51       5.82%
3         group B                            college  ...     54       6.17%
4         group B                       postgraduate  ...      5       0.57%
5         group B                             school  ...    107       12.2%
6         group C                            college  ...    102       11.6%
7         group C                       postgraduate  ...     15       1.71%
8         group C                             school  ...    155       17.7%
9         group D                            college  ...     70        8.0%
10        group D                       postgraduate  ...     20       2.28%
11        group D                             school  ...    149       17.0%
12        group E                            college  ...     52       5.94%
13        group E                       postgraduate  ...      6       0.68%
14        group E                             school  ...     63        7.2%

[15 rows x 8 columns]

**Resultado Esperado**

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.74 |            5    |            4.89 |  4.88 |      24 | 2.74 %       |
|  1 |           | postgraduate                         |         4.69 |            5.23 |            5.35 |  5.09 |       2 | 0.23 %       |
|  2 |           | school                               |         4.57 |            4.73 |            4.56 |  4.62 |      51 | 5.83 %       |
|  3 | group B          | college                              |         5.07 |            5.26 |            5.19 |  5.18 |      54 | 6.17 %       |
|  4 |           | postgraduate                         |         4.91 |            5.69 |            5.55 |  5.38 |       5 | 0.57 %       |
|  5 |           | school                               |         4.69 |            4.89 |            4.76 |  4.78 |     107 | 12.23 %      |
|  6 | group C          | college                              |         5.02 |            5.37 |            5.35 |  5.25 |     102 | 11.66 %      |
|  7 |           | postgraduate                         |         4.92 |            5.14 |            5.1  |  5.06 |      15 | 1.71 %       |
|  8 |           | school                               |         4.76 |            5.02 |            4.92 |  4.9  |     155 | 17.71 %      |
|  9 | group D          | college                              |         5.11 |            5.25 |            5.25 |  5.2  |      70 | 8.0 %        |
| 10 |           | postgraduate                         |         5.22 |            5.54 |            5.73 |  5.5  |      20 | 2.29 %       |
| 11 |           | school                               |         5.02 |            5.13 |            5.11 |  5.09 |     149 | 17.03 %      |
| 12 | group E          | college                              |         5.54 |            5.45 |            5.45 |  5.48 |      52 | 5.94 %       |
| 13 |          | postgraduate                         |         5.54 |            6.03 |            5.89 |  5.82 |       6 | 0.69 %       |
| 14 |          | school                               |         5.4  |            5.31 |            5.16 |  5.29 |      63 | 7.2 %        |

## Visualizaciones [0.5 Puntos]

Ahora, implemente un gráfico de caja en donde se muestre el GPA con respecto al nivel educacional y que la variable de color sea la etnicidad y luego comente.

In [ ]:
def plot_box_etnia(df):
  

plot_box_etnia(df_grades)

In [176]:
fig = px.box(df_grades, x="simple parental level of education", y="GPA")
fig.show()

> 1. ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?
> 2. ¿Este gráfico permite hacer facilmente un análisis conjunto de estas dos variables de forma sencilla?

**Justifique:**

```
Escriba aqui su respuesta
```

# 1.3.5 Combinar Dataset [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias. 
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene nuevamente los alumnos de los registros corruptos que ud. y su equipo filtraron en el análisis anterior.
 
 
El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`). 


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados).

**To-Do**

- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

In [ ]:
...

> **Justificación:**


```
Escriba aquí su justificación
```

## 1.4 Más visualizaciones [Bonus - 0.5 puntos por cada Gráfico, máximo 1 punto]


**Desde aquí parte el bonus para el lab.**


<p align="center">
  <img src="https://media4.giphy.com/media/xT5LMWNOjGqJzUfyve/giphy.gif?cid=ecf05e47phkhvg6vadqvcoobvxwj9bzxxk38jjvkcc6eak6t&rid=giphy.gif&ct=g">
</p>




Genere dos visualizaciones extras que encuentre interesantes (y no triviales) con estos datos y explique sus resultados. Agrupe los atributos que estime convenientes.


**To-Do:**
- [ ] Generar dos nuevas visualizaciones con los datos y explicar que están representando.


> **NOTA: No utilice historia ni ciencias, son notas generadas aleatoriamente.**

# Conclusión
Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>

<br>
<center>
<img src="https://i.kym-cdn.com/photos/images/original/001/194/195/b18.png" width=100 height=50 />
</center>
<br>